File: windows/Tokendial.App/Panel/Theme.cs
using System.Windows;
using System.Windows.Media;
using System.Windows.Media.Animation;
using Tokendial.Core.Model;

namespace Tokendial.App.Panel;

/// <summary>docs/design/tokens.md, as WPF resources. Frozen brushes; every number in device-independent points.</summary>
public static class Theme
{
    public static readonly Brush Surface = Freeze(new SolidColorBrush(Color.FromArgb(128, 16, 17, 20)));
    public static readonly Brush SurfaceEdge = Freeze(new SolidColorBrush(Color.FromArgb(20, 255, 255, 255)));
    public static readonly Brush CardSurface = Freeze(new SolidColorBrush(Color.FromArgb(235, 16, 17, 20)));
    public static readonly Brush Track = Freeze(new SolidColorBrush(Color.FromArgb(41, 255, 255, 255)));
    public static readonly Brush Ample = Freeze(new SolidColorBrush(Color.FromRgb(0x34, 0xD3, 0x99)));
    public static readonly Brush Watch = Freeze(new SolidColorBrush(Color.FromRgb(0xFB, 0xBF, 0x24)));
    public static readonly Brush Critical = Freeze(new SolidColorBrush(Color.FromRgb(0xFB, 0x71, 0x85)));
    public static readonly Brush Working = Freeze(new SolidColorBrush(Color.FromRgb(0xF5, 0xF5, 0xF7)));
    public static readonly Brush Waiting = Watch;
    public static readonly Brush TextPrimary = Freeze(new SolidColorBrush(Color.FromRgb(0xF5, 0xF5, 0xF7)));
    public static readonly Brush TextSecondary = Freeze(new SolidColorBrush(Color.FromRgb(0x9A, 0x9D, 0xA6)));
    public static readonly Brush TextDisabled = Freeze(new SolidColorBrush(Color.FromRgb(0x5C, 0x5F, 0x68)));
    public static readonly Brush Hairline = Freeze(new SolidColorBrush(Color.FromArgb(28, 255, 255, 255)));
    public static readonly Brush Transparent = Brushes.Transparent;
    public static readonly Color AmpleColor = Color.FromRgb(0x34, 0xD3, 0x99);
    public static readonly Color TextDisabledColor = Color.FromRgb(0x5C, 0x5F, 0x68);

    public static readonly FontFamily Font = new("Segoe UI Variable Text, Segoe UI Variable, Segoe UI");
    public static readonly FontFamily DisplayFont = new("Segoe UI Variable Display, Segoe UI Variable, Segoe UI");

    public const double CompactHeight = 34;
    public const double CompactPadding = 14;
    public const double CompactSpacing = 10;
    public const double CompactDial = 24;
    public const double CompactMark = 11;
    public const double ExpandedMark = 13;
    public const double CompactStroke = 3;
    public const double ExpandedHeight = 132;
    public const double ExpandedPadding = 20;
    public const double CellWidth = 96;
    public const double ExpandedDial = 56;
    public const double ExpandedStroke = 6;
    public const double ActivityDial = 40;
    public const double ActivityStroke = 2;
    public const double CompactRadius = 17;
    public const double ExpandedRadius = 20;
    public const double HotZone = 24;
    public const double DockSlant = 14;
    public const double CardWidth = 280;
    public const double CardRadius = 16;
    public const double CardPadding = 14;

    public static Brush Of(Band? band) => band switch
    {
        Band.Ample => Ample,
        Band.Watch => Watch,
        Band.Critical => Critical,
        _ => TextDisabled
    };

    public static Brush Of(double? fraction) => fraction is double f ? Of(Bands.Of(f)) : TextDisabled;

    public static readonly SpringCurve Expand = new(0.38, 0.80);
    public static readonly SpringCurve Contents = new(0.32, 0.84);
    public static readonly SpringCurve Reading = new(0.80, 0.90);
    public static readonly SpringCurve Glide = new(0.45, 0.86);
    public static readonly Duration Crossfade = new(TimeSpan.FromSeconds(0.15));

    public static bool ReduceMotion => !SystemParameters.ClientAreaAnimation;

    public static Duration DurationOf(SpringCurve curve) => ReduceMotion ? new Duration(TimeSpan.Zero) : new Duration(curve.SettleTime);

    public static double Stagger(int index) => ReduceMotion ? 0 : Math.Min(index * 0.04, 0.16);

    private static Brush Freeze(Brush brush)
    {
        brush.Freeze();
        return brush;
    }
}

/// <summary>A critically- or under-damped spring sampled over the time it takes to settle within a thousandth.</summary>
public sealed class SpringCurve : EasingFunctionBase
{
    private readonly double omega;
    private readonly double zeta;

    public SpringCurve(double response, double damping)
    {
        Response = response;
        Damping = damping;
        omega = 2 * Math.PI / response;
        zeta = damping;
        SettleTime = TimeSpan.FromSeconds(Math.Log(1000) / (zeta * omega));
        EasingMode = EasingMode.EaseIn;
    }

    public double Response { get; }
    public double Damping { get; }
    public TimeSpan SettleTime { get; }

    protected override double EaseInCore(double normalizedTime)
    {
        var t = normalizedTime * SettleTime.TotalSeconds;
        var decay = Math.Exp(-zeta * omega * t);
        if (zeta >= 1) return 1 - decay * (1 + omega * t);
        var damped = omega * Math.Sqrt(1 - zeta * zeta);
        return 1 - decay * (Math.Cos(damped * t) + zeta * omega / damped * Math.Sin(damped * t));
    }

    protected override Freezable CreateInstanceCore() => new SpringCurve(Response, Damping);
}

File: macos/Tokendial/Theme.swift
import AppKit
import SwiftUI
import TokendialCore

/// docs/design/tokens.md as AppKit values. Every number in points; every colour comes from the palette in use.
enum Theme {
    /// One look for the dock and the card: surfaces, text inks and the band colours.
    struct Palette {
        let surface: NSColor
        let surfaceEdge: NSColor
        let cardSurface: NSColor
        let track: NSColor
        let ample: NSColor
        let watch: NSColor
        let critical: NSColor
        let working: NSColor
        let textPrimary: NSColor
        let textSecondary: NSColor
        let textDisabled: NSColor
        let hairline: NSColor
    }

    static let darkPalette = Palette(
        surface: rgba(16, 17, 20, 0.50), surfaceEdge: white(1, 0.08), cardSurface: rgba(16, 17, 20, 0.92), track: white(1, 0.16),
        ample: rgb(0x34, 0xD3, 0x99), watch: rgb(0xFB, 0xBF, 0x24), critical: rgb(0xFB, 0x71, 0x85), working: rgb(0xF5, 0xF5, 0xF7),
        textPrimary: rgb(0xF5, 0xF5, 0xF7), textSecondary: rgb(0x9A, 0x9D, 0xA6), textDisabled: rgb(0x5C, 0x5F, 0x68), hairline: white(1, 0.11))

    static let lightPalette = Palette(
        surface: rgba(248, 249, 251, 0.80), surfaceEdge: rgba(15, 23, 42, 0.10), cardSurface: rgba(252, 252, 253, 0.96), track: rgba(15, 23, 42, 0.12),
        ample: rgb(0x10, 0xB9, 0x81), watch: rgb(0xF5, 0x9E, 0x0B), critical: rgb(0xF4, 0x3F, 0x5E), working: rgb(0x1B, 0x1F, 0x27),
        textPrimary: rgb(0x1B, 0x1F, 0x27), textSecondary: rgb(0x5B, 0x62, 0x70), textDisabled: rgb(0xA3, 0xA9, 0xB4), hairline: rgba(15, 23, 42, 0.10))

    private(set) static var dark = true
    private static var palette = darkPalette

    /// Switch the look. Views built afterwards pick it up; live ones are repainted by their windows.
    static func use(dark isDark: Bool) {
        dark = isDark
        palette = isDark ? darkPalette : lightPalette
    }

    static var surface: NSColor { palette.surface }
    static var cardSurface: NSColor { palette.cardSurface }
    static var surfaceEdge: NSColor { palette.surfaceEdge }
    static var track: NSColor { palette.track }
    static var ample: NSColor { palette.ample }
    static var watch: NSColor { palette.watch }
    static var critical: NSColor { palette.critical }
    static var working: NSColor { palette.working }
    static var waiting: NSColor { palette.watch }
    static var textPrimary: NSColor { palette.textPrimary }
    static var textSecondary: NSColor { palette.textSecondary }
    static var textDisabled: NSColor { palette.textDisabled }
    static var hairline: NSColor { palette.hairline }

    static let compactHeight: CGFloat = 28
    static let compactPadding: CGFloat = 14
    static let compactSpacing: CGFloat = 10
    static let compactDial: CGFloat = 18
    static let compactStroke: CGFloat = 3
    static let expandedHeight: CGFloat = 132
    static let expandedPadding: CGFloat = 20
    static let cellWidth: CGFloat = 96
    static let expandedDial: CGFloat = 56
    static let expandedStroke: CGFloat = 6
    static let activityDial: CGFloat = 40
    static let activityStroke: CGFloat = 2
    static let slant: CGFloat = 14
    static let compactMark: CGFloat = 11
    static let expandedMark: CGFloat = 13
    static let compactRadius: CGFloat = 14
    static let expandedRadius: CGFloat = 20
    static let hotZone: CGFloat = 24
    static let cardWidth: CGFloat = 280
    static let cardRadius: CGFloat = 16
    static let cardPadding: CGFloat = 14

    static func color(_ band: Band?) -> NSColor {
        switch band { case .ample?: return ample; case .watch?: return watch; case .critical?: return critical; case nil: return textDisabled }
    }

    static func color(fraction: Double?) -> NSColor { fraction.map { color(Band.of($0)) } ?? textDisabled }

    static func font(_ size: CGFloat, weight: NSFont.Weight = .regular) -> NSFont {
        NSFont.monospacedDigitSystemFont(ofSize: size, weight: weight)
    }

    static var reduceMotion: Bool { NSWorkspace.shared.accessibilityDisplayShouldReduceMotion }

    static func compactWidth(_ n: Int) -> CGFloat { n == 0 ? 2 * compactPadding + 40 : 2 * compactPadding + CGFloat(n) * compactDial + CGFloat(n - 1) * compactSpacing }
    static func expandedWidth(_ n: Int) -> CGFloat { max(2 * expandedPadding + CGFloat(max(n, 1)) * cellWidth, cardWidth + 2 * expandedPadding) }

    private static func rgb(_ r: Int, _ g: Int, _ b: Int) -> NSColor {
        NSColor(srgbRed: CGFloat(r) / 255, green: CGFloat(g) / 255, blue: CGFloat(b) / 255, alpha: 1)
    }

    private static func rgba(_ r: Int, _ g: Int, _ b: Int, _ alpha: CGFloat) -> NSColor {
        NSColor(srgbRed: CGFloat(r) / 255, green: CGFloat(g) / 255, blue: CGFloat(b) / 255, alpha: alpha)
    }

    private static func white(_ level: CGFloat, _ alpha: CGFloat) -> NSColor { NSColor(white: level, alpha: alpha) }
}

/// The same tokens as SwiftUI colours, for the settings and welcome windows.
extension Theme {
    enum UI {
        static var surface: Color { Color(nsColor: Theme.surface) }
        static var cardSurface: Color { Color(nsColor: Theme.cardSurface) }
        static var surfaceEdge: Color { Color(nsColor: Theme.surfaceEdge) }
        static var track: Color { Color(nsColor: Theme.track) }
        static var ample: Color { Color(nsColor: Theme.ample) }
        static var watch: Color { Color(nsColor: Theme.watch) }
        static var critical: Color { Color(nsColor: Theme.critical) }
        static var textPrimary: Color { Color(nsColor: Theme.textPrimary) }
        static var textSecondary: Color { Color(nsColor: Theme.textSecondary) }
        static var textDisabled: Color { Color(nsColor: Theme.textDisabled) }
        static var hairline: Color { Color(nsColor: Theme.hairline) }
    }
}

/// Springs from the tokens as Core Animation springs.
struct Spring {
    let response: Double
    let damping: Double

    static let expand = Spring(response: 0.38, damping: 0.80)
    static let contents = Spring(response: 0.32, damping: 0.84)
    static let reading = Spring(response: 0.80, damping: 0.90)
    static let glide = Spring(response: 0.45, damping: 0.86)

    var omega: Double { 2 * .pi / response }
    var settle: TimeInterval { log(1000) / (damping * omega) }

    func animation(_ keyPath: String, from: Any?, to: Any) -> CAAnimation {
        let spring = CASpringAnimation(keyPath: keyPath)
        spring.mass = 1
        spring.stiffness = omega * omega
        spring.damping = 2 * damping * omega
        spring.fromValue = from
        spring.toValue = to
        spring.duration = spring.settlingDuration
        spring.fillMode = .forwards
        spring.isRemovedOnCompletion = false
        return spring
    }
}
